### PR TITLE
Add Declare Exchange, Quque

### PR DESCRIPTION
**Pull request by dshvedchenko**
Binding implemented (!) targets have to exists.
Durability, Autodelete, Exclusiveness - for Queue in Consumer Mode
Durability, Autodelete, Exclusiveness - for Exchange in Producer Mode

### CODE REVIEW

File: src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java [hunk 1]
```diff
@@ -298,14 +304,21 @@ private void initFilter() throws Exception
...
+        String vhost      = environmentSubstitute(meta.getVhost());
+
 
         if (body == null) {
             throw new AMQPException("Unable to retrieve field : " + meta.getBodyField());
         }
 
-        if (uri == null) {
-            throw new AMQPException("Unable to retrieve connection uri");
+        if (username == null || password == null || host == null || port == null || vhost == null) {
```
Comment (FabioBatSilva) [inline src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java]:
``` java
if ((username == null || password == null || host == null || port == null) && (uri == null) ) {
    throw new AMQPException("Unable to retrieve connection information");
}
```

File: src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java [hunk 2]
```diff
@@ -333,19 +365,40 @@ private void initFilter() throws Exception
...
         logMinimal(getString("AmqpPlugin.Body.Label")      + " : " + body);
         logMinimal(getString("AmqpPlugin.Routing.Label")   + " : " + routing);
         logMinimal(getString("AmqpPlugin.Target.Label")    + " : " + data.target);
         logMinimal(getString("AmqpPlugin.Limit.Label")     + " : " + data.limit);
 
-        factory.setUri(uri);
+
+	if (uri != null && !Const.isEmpty(uri)) {
+	   factory.setUri(uri);
+	   String uri_passwd_masked = uri.replace(factory.getPassword(),"*****");
```
Comment (FabioBatSilva) [inline src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java]:
``` java
String maskedPassword
```

Comment (FabioBatSilva) [inline src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java]:
``` java
if (vhost != null) {
    factory.setVirtualHost(vhost);
}
```

File: src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPluginDialog.java [hunk 3]
```diff
@@ -463,6 +910,24 @@ private void ok()
             input.setLimit(textLimit.getText());
         }
 
+
+	//Save Binding Table
+	int count = wBinding.nrNonEmpty();
+	// in Cosumer mode if we declare queu , we have to bind it to Target
+	if ( checkDeclare.getSelection() && count ==  0 && input.getMode() == AMQPPluginData.MODE_CONSUMER ) {
```
Comment (FabioBatSilva) [inline src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPluginDialog.java]:
``` java
 AMQPPluginData.MODE_CONSUMER.equals(input.getMode())
```

File: src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPluginData.java [hunk 4]
```diff
@@ -23,6 +34,18 @@
     public Long limit;
     public long count = 0;
 
+
```
Comment (FabioBatSilva) [inline src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPluginData.java:39]: Can you make this something like this ?: 

``` java
public class AMQPPluginMeta
{
    // ....
    class Binding
    {
        private String target[];
        private String type[];
        private String routing[];
    }

    public void setBindingRoutingValue(Binding[] bindings)
    {
        this.bindings = bindings;
    }
}
```

**Comment by FabioBatSilva** [comment]
Some times I'm getting a "unable to open dialog for this step : -1" exception while editing the binding  table, probably related to the iteration over the table items

**Comment by FabioBatSilva** [comment]
Nice work @dshvedchenko,
Let me know when you have this minor fixes done and i'll take a look again

File: src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPluginData.java (hunk 4 above)
Comment (dshvedchenko) [inline src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPluginData.java:39]: of course

**Comment by dshvedchenko** [comment]
(!) for now I removed declaration for targets. Declaration for Queue/Exchange itself for Consumer/producer mode, and Binding.
@FabioBatSilva what do you think about implementing declaration for Target ? is it necessary ?
it will take extra 2 - 3 hours, but will it used by users ?

**Comment by FabioBatSilva** [comment]
Its i nice to have, Queue/Exchange declaration will handle most of the cases.
Anyone how really need the "target" declaration can add another step before it to declare the "target".

File: src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPluginData.java (hunk 4 above)
Comment (FabioBatSilva) [inline src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPluginData.java:39]: If we make bindings a list we can get rid of the allocate method..

``` java
private List<Binding> bindings = new ArrayList<Binding>();

public void addBinding(String target, String routing)
{
    this.bindings.add(new Binding(target, routing));
}

public List<Binding> getBindings()
{
    return this.bindings;
}
```

**Comment by dshvedchenko** [comment]
BTW: how to treat current binding, should we remove already defined before add new ( defined in kettle ) ?

I think we at least can introduce auto-unbind on transformation finish, but it should be another issue

**Comment by dshvedchenko** [comment]
ArrayList used

**Comment by FabioBatSilva** [comment]
I believe the current behaviour will work fine..

IIRC Declaring simply ensures that it exists, creating it if necessary.
So if you redeclare a queue/exchange/bind it should be no problem.

And If the user don`t need the queue/bind after the the kettle run he can declare it as not durable and rabbit will delete it after the transformation finish and disconnect.

**Comment by dshvedchenko** [comment]
currently consumer implemented, so if queue is empty at the moment transformation start, it will process nothing.

should we add mode SUBSCRIBER, to wait for some number of messages ??? maybe another issue ?

**Comment by FabioBatSilva** [comment]
We can open a issue to discuss this, 
But It can probably be done as part of the transformation

Something like :

```
"Declare Queue/Bind Step" ->  "Wait for X Sec Step" -> "Queue Consume Step"
```

Also transformation/jobs like that are usually scheduled to run from time to time,
So in some cases its OK if the queue is empty in the first run the messages will be consumed durrring the next run.

**Comment by FabioBatSilva** [comment]
@dshvedchenko I'll do another test run and merge it you are done with your changes..

**Comment by dshvedchenko** [comment]
ok, so I'm waiting for you comments on current implementation and final decision about merge
thank you

**Comment by FabioBatSilva** [comment]
Cool, i'll merge soon...

**Comment by dshvedchenko** [comment]
maybe we need extend text for limitation in Readme.md ...

**Comment by dbron0000** [comment]
That would be nice for me. I have that function now but it is quite inefficient. A more thoughtful approach would be great for a subscriber,

**Comment by FabioBatSilva** [comment]
Nice work @dshvedchenko

**Pull request merged**